planet/clients/orders.py

```diff
@@ -149,30 +149,23 @@ async def create_order(
             request: order request definition
 
         Returns:
-            The ID of the order
+            Order
 
         Raises:
-            planet.exceptions.APIException: On API error.
-        '''
+            OrderError if order creation fails. The exception will
+            usually have a __context__ with a very verbose error
+            messaage from the service endpoint.
+
+        """
         url = self._orders_url()
         req = self._request(url, method='POST', json=request)
 
         try:
             resp = await self._do_request(req)
-        except exceptions.BadQuery as ex:
-            msg_json = json.loads(ex.message)
-
-            msg = msg_json['general'][0]['message']
-            try:
-                # get first error field
-                field = next(iter(msg_json['field'].keys()))
-                msg += ' - ' + msg_json['field'][field][0]['message']
-            except AttributeError:
-                pass
-            raise exceptions.BadQuery(msg)
+        except exceptions.BadQuery as bad_query:
+            raise OrderError("Order creation failed.") from bad_query
```

Comment:
Here's the gist of it: for the Python API we chain the exceptions. The user will get an OrderError with a brief, unambiguous message and this carries the BadQuery exception with it. And then we pass the responsibility for presenting the BadQuery message to a human onto the caller of create_order. Advantage for the Python API is that we generalize our exception so that it can suit the needs of almost any caller and eliminate code. Advantage for a user is that they are not bound by choices we make.

Reply:
@jreiberkyle what do you think about hiding the request/response details and raising a more general OrderError for create_order?

Reply:
I am going back and forth here quite a bit. I like the clarity that adding the `OrderError` layer brings in that it can specify that the order creation failed. However, it seems like we are handing the user quite a bit of complication if we are telling them to use `error.__context__.message` to access the real source of the failure. When the two usages are compared, I have to say that the second one seems easier. And there are reasons besides API interactions that can cause a failed order - like, for example, if some of the user input wasn't quite right.

```python
try:
    order = client.create_order(...)
except OrderError as error:
    print("Order creation failed. Service response details: {}".format(error.__context__.message))
```

```python
try:
    order = client.create_order(...)
except APIException as exception:
   print("Order creation failed. Service response details: {}".format(exception.message))
```

Reply:
Yes, it's very possible that users will want easier access to the details about the root cause of a server error. I propose we deal with that later and for now go with the chaining, which doesn't lose any information from the server and provides a good base for future improvements.